File: scale_drag_recyclerview/src/main/java/android/support/v7/widget/CenterLinearSnapHelper.java
package android.support.v7.widget;

import android.graphics.PointF;
import android.graphics.Rect;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.OrientationHelper;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SnapHelper;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

public class CenterLinearSnapHelper extends LinearSnapHelper {
    @Nullable
    private OrientationHelper mVerticalHelper;
    @Nullable
    private OrientationHelper mHorizontalHelper;

    @Override
    public View findSnapView(RecyclerView.LayoutManager layoutManager) {
        if (layoutManager.canScrollVertically()) {
            return this.findCenterView(layoutManager, this.getVerticalHelper(layoutManager));
        } else {
            return layoutManager.canScrollHorizontally() ? this.findCenterView(layoutManager, this.getHorizontalHelper(layoutManager)) : null;
        }
    }

    @Nullable
    private View findCenterView(RecyclerView.LayoutManager layoutManager, OrientationHelper helper) {
        int childCount = layoutManager.getChildCount();
        if (childCount == 0) {
            return null;
        } else {
            View closestChild = null;
            int center;
            if (layoutManager.getClipToPadding()) {
                center = helper.getStartAfterPadding() + helper.getTotalSpace() / 2;
            } else {
                center = helper.getEnd() / 2;
            }

            int absClosest = Integer.MAX_VALUE;
            for (int i = 0; i < childCount; ++i) {
                View child = layoutManager.getChildAt(i);

                // int childCenter = helper.getDecoratedStart(child) + helper.getDecoratedMeasurement(child) / 2;

                //---- start -----
                RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
                Rect insets = params.mDecorInsets;
                int childCenter;
                if (layoutManager.canScrollVertically()) {
                    childCenter = helper.getDecoratedStart(child) + insets.top + params.topMargin + child.getMeasuredHeight() / 2;
                } else {
                    childCenter = helper.getDecoratedStart(child) + insets.left + params.leftMargin + child.getMeasuredWidth() / 2;
                }
                //--- end ----

                int absDistance = Math.abs(childCenter - center);
                if (absDistance < absClosest) {
                    absClosest = absDistance;
                    closestChild = child;
                }
            }

            return closestChild;
        }
    }

    @NonNull
    private OrientationHelper getVerticalHelper(@NonNull RecyclerView.LayoutManager layoutManager) {
        if (this.mVerticalHelper == null || this.mVerticalHelper.mLayoutManager != layoutManager) {
            this.mVerticalHelper = OrientationHelper.createVerticalHelper(layoutManager);
        }

        return this.mVerticalHelper;
    }

    @NonNull
    private OrientationHelper getHorizontalHelper(@NonNull RecyclerView.LayoutManager layoutManager) {
        if (this.mHorizontalHelper == null || this.mHorizontalHelper.mLayoutManager != layoutManager) {
            this.mHorizontalHelper = OrientationHelper.createHorizontalHelper(layoutManager);
        }

        return this.mHorizontalHelper;
    }
}

File: scale_drag_recyclerview/src/main/java/com/frisky/view/scalerecyclerview/CenterLinearSnapHelper.kt
package com.frisky.view.scalerecyclerview

import android.view.View
import androidx.recyclerview.widget.LinearSnapHelper
import androidx.recyclerview.widget.OrientationHelper
import androidx.recyclerview.widget.RecyclerView

class CenterLinearSnapHelper : LinearSnapHelper() {
    private var mVerticalHelper: OrientationHelper? = null
    private var mHorizontalHelper: OrientationHelper? = null

    override fun findSnapView(layoutManager: RecyclerView.LayoutManager): View? {
        return if (layoutManager.canScrollVertically()) {
            findCenterView(layoutManager, getVerticalHelper(layoutManager))
        } else if (layoutManager.canScrollHorizontally()) {
            findCenterView(layoutManager, getHorizontalHelper(layoutManager))
        } else {
            null
        }
    }

    private fun findCenterView(layoutManager: RecyclerView.LayoutManager, helper: OrientationHelper): View? {
        val childCount = layoutManager.childCount
        if (childCount == 0) {
            return null
        } else {
            var closestChild: View? = null
            val center = if (layoutManager.clipToPadding) {
                helper.startAfterPadding + helper.totalSpace / 2
            } else {
                helper.end / 2
            }
            var absClosest = Int.MAX_VALUE
            for (i in 0 until childCount) {
                val child = layoutManager.getChildAt(i) ?: continue

                val params = child.layoutParams as RecyclerView.LayoutParams
                val childCenter = if (layoutManager.canScrollVertically()) {
                    helper.getDecoratedStart(child) + layoutManager.getTopDecorationHeight(child) + params.topMargin + child.measuredHeight / 2
                } else {
                    helper.getDecoratedStart(child) + layoutManager.getLeftDecorationWidth(child) + params.leftMargin + child.measuredWidth / 2
                }

                val absDistance = Math.abs(childCenter - center)
                if (absDistance < absClosest) {
                    absClosest = absDistance
                    closestChild = child
                }
            }
            return closestChild
        }
    }

    private fun getVerticalHelper(layoutManager: RecyclerView.LayoutManager): OrientationHelper {
        if (this.mVerticalHelper == null || this.mVerticalHelper!!.layoutManager !== layoutManager) {
            this.mVerticalHelper = OrientationHelper.createVerticalHelper(layoutManager)
        }
        return this.mVerticalHelper!!
    }

    private fun getHorizontalHelper(layoutManager: RecyclerView.LayoutManager): OrientationHelper {
        if (this.mHorizontalHelper == null || this.mHorizontalHelper!!.layoutManager !== layoutManager) {
            this.mHorizontalHelper = OrientationHelper.createHorizontalHelper(layoutManager)
        }
        return this.mHorizontalHelper!!
    }
}